### DISCUSSION TITLE
Modifying the training graph in java

### DISCUSSION THREAD

**Discussion by heflinstephenraj** [microsoft/onnxruntime#19649] [API Q&A] [answered]
Hi,

I am training a model in Java using ONNX. I have exported the training graph, evaluation graph, optimizer graph, and checkpoint in Python. I used these in ONNX Java to train the model. Because these graphs are fixed in Java, I can't change the model parameters when my data has changed.

For example, if I export the graph with 10 neurons in the output layer, I can train the model in Java with data that has 10 labels. If my dataset has more than 10 labels, then again I should go back to Python and generate the graph. 
On the other hand, if my graph has 10 neurons in the output layer and my data has only 5 labels, I am wasting computation on those unused 5 neurons in the output layer.

Is there any possibility to modify the graph in Java?

**Comment by Craigacp** [comment, marked as answer]
Not at the moment. The code which generates the different graphs based on the original ONNX model only has a Python interface. If it becomes a supported C or C++ API we could wrap a Java binding around it, and then construct the ONNX models from Java using something like Tribuo's [ONNX generation support](https://github.com/oracle/tribuo/tree/main/Util/ONNXExport/src/main/java/org/tribuo/util/onnx), though Tribuo's ONNX generation support doesn't cover all ONNX ops yet, or we could write the initial model protobuf directly.
